### PR TITLE
dns: frame support - v3

### PR DESCRIPTION
**Pull request by jasonish**
- frames (rust): method to create frames
- dns: add pdu frame

Ticket: https://redmine.openinfosecfoundation.org/issues/4984

### CODE REVIEW

File: rust/src/dns/dns.rs [hunk 1]
```diff
@@ -492,8 +519,11 @@ impl DNSState {
             SCLogDebug!("[request] Have {} bytes, need {} to parse",
                         cur_i.len(), size + 2);
             if size > 0 && cur_i.len() >= size + 2 {
-                let msg = &cur_i[0..(size + 2)];
-                if self.parse_request(&msg[2..]) {
+                let msg = &cur_i[2..(size + 2)];
+                if cfg!(not(test)) {
```
Comment (victorjulien) [inline rust/src/dns/dns.rs:523]: is there a more graceful way to handle this? Ideally we'd not need it at all

Comment (jasonish) [reply rust/src/dns/dns.rs:523]: > is there a more graceful way to handle this? Ideally we'd not need it at all

We could take the C functions in this code path and convert them to function pointers set via `SuricataContext` then make these calls no-ops when called from tests.  Not  sure that qualifies as more graceful though.

Comment (victorjulien) [reply rust/src/dns/dns.rs:523]: Can we hide the `if cfg!(not(test)) {` inside the `Frame` implementation somehow, so it won't pollute the parser logic?

Comment (jasonish) [reply rust/src/dns/dns.rs:523]: Yeah, makes sense.

**Comment by jasonish** [comment]
https://github.com/OISF/suricata/pull/7316

**Pull request closed without merging**